src/common/FastRational.h

```diff
@@ -539,21 +551,34 @@ inline int cmpabs(FastRational op1, FastRational op2)
 };
 template<ulword> ulword gcd(ulword a, ulword b);
 template<uword> uword gcd(uword a, uword b);
-#define CHECK_WORD(var, value) \
-    do { \
-        lword tmp = value; \
-        if (tmp < WORD_MIN/2 || tmp > WORD_MAX/2) { \
-            goto overflow; \
-        } \
-        var = tmp;\
-    } while(0)
+#define CHECK_WORD(var, value)                  \
+    do {                                        \
+        lword tmp = value;                      \
+        if (tmp < WORD_MIN || tmp > WORD_MAX) { \
+            goto overflow;                      \
+        }                                       \
+        var = tmp;                              \
+    } while(0)                                  \
+
+// Adapted from https://codereview.stackexchange.com/questions/37177/simple-method-to-detect-int-overflow
 #define CHECK_SUM_OVERFLOWS_LWORD(var, s1, s2) \
-    do { \
-        if ((s1 > LWORD_MAX/2 || s2 > LWORD_MAX/2) || (s1 < LWORD_MIN/2 || s2 < LWORD_MIN/2)) { \
+    do {                                       \
+        if (s1 > 0 and s2 > LWORD_MAX - s1) {  \
+            goto overflow;                     \
+        } if (s1 < 0 and s2 < LWORD_MIN - s1) {\
+            goto overflow;                     \
+        }                                      \
+        var = s1 + s2;                         \
+    } while (0)                                \
+
+#define CHECK_SUB_OVERFLOWS_LWORD(var, s1, s2) \
+    do {                                       \
+        if ((s1 > LWORD_MAX/2 || s2 < LWORD_MIN/2) || (s1 < LWORD_MIN/2 || s2 > LWORD_MAX/2)) { \
             goto overflow; \
         } \
-        var = s1 + s2;\
-    } while(0)
+        var = s1 - s2; \
```

Comment:
There should be a better way of doing this check (see `CHECK_SUM_OVERFLOWS_LWORD`)?